std/encoding.d

```diff
@@ -3380,3 +3380,111 @@ version(unittest)
+    utf32be   = 1,  /// [0x00, 0x00, 0xFE, 0xFF]
+    utf32le   = 2,  /// [0xFF, 0xFE, 0x00, 0x00]
+    utf7      = 3,  /*  [0x2B, 0x2F, 0x76, 0x38]
+                        [0x2B, 0x2F, 0x76, 0x39],
+                        [0x2B, 0x2F, 0x76, 0x2B],
+                        [0x2B, 0x2F, 0x76, 0x2F],
+                        [0x2B, 0x2F, 0x76, 0x38, 0x2D]
+                    */
+    utf1      = 8,  /// [0xF7, 0x64, 0x4C]
+    utfebcdic = 9,  /// [0xDD, 0x73, 0x66, 0x73]
+    scsu      = 10,  /// [0x0E, 0xFE, 0xFF]
+    bocu1     = 11, /// [0xFB, 0xEE, 0x28]
+    gb18030   = 12, /// [0x84, 0x31, 0x95, 0x33]
+    utf8      = 13, /// [0xEF, 0xBB, 0xBF]
+    utf16be   = 14, /// [0xFE, 0xFF]
+    utf16le   = 15  /// [0xFF, 0xFE]
+}
+
+/// The type stored inside $(D bomTable).
+alias BOMSeq = Tuple!(BOM, "bom", ubyte[], "seq");
+
+/** Mapping of a byte sequence to $(B Byte Order Mark (BOM))
+*/
+immutable bomTable = [
+        BOMSeq(BOM.noBOM, new ubyte[0]),
+        BOMSeq(BOM.utf32be, to!(ubyte[])([0x00, 0x00, 0xFE, 0xFF])),
+        BOMSeq(BOM.utf32le, to!(ubyte[])([0xFF, 0xFE, 0x00, 0x00])),
+        BOMSeq(BOM.utf7, to!(ubyte[])([0x2B, 0x2F, 0x76, 0x39])),
+        BOMSeq(BOM.utf7, to!(ubyte[])([0x2B, 0x2F, 0x76, 0x2B])),
+        BOMSeq(BOM.utf7, to!(ubyte[])([0x2B, 0x2F, 0x76, 0x2F])),
+        BOMSeq(BOM.utf7, to!(ubyte[])([0x2B, 0x2F, 0x76, 0x38, 0x2D])),
+        BOMSeq(BOM.utf7, to!(ubyte[])([0x2B, 0x2F, 0x76, 0x38])),
+        BOMSeq(BOM.utf1, to!(ubyte[])([0xF7, 0x64, 0x4C])),
+        BOMSeq(BOM.utfebcdic, to!(ubyte[])([0xDD, 0x73, 0x66, 0x73])),
+        BOMSeq(BOM.scsu, to!(ubyte[])([0x0E, 0xFE, 0xFF])),
+        BOMSeq(BOM.bocu1, to!(ubyte[])([0xFB, 0xEE, 0x28])),
+        BOMSeq(BOM.gb18030, to!(ubyte[])([0x84, 0x31, 0x95, 0x33])),
+        BOMSeq(BOM.utf8, to!(ubyte[])([0xEF, 0xBB, 0xBF])),
+        BOMSeq(BOM.utf16be, to!(ubyte[])([0xFE, 0xFF])),
+        BOMSeq(BOM.utf16le, to!(ubyte[])([0xFF, 0xFE]))
+];
+
+/** Returns a $(D Tuple!(BOM,ubyte[])) for a given $(D input).
+If no $(D BOM) is present the $(D Tuple!(BOM,ubyte[]) for $(D BOM.noBOM) is
+returned.
+
+Params:
+    input = The sequence to check for the $(D BOM)
+
+Returns:
+    the found $(D Tuple!(BOM,ubyte[])) of $(D BOM) and the $(D BOM) sequence
+    as a $(D ubyte[]).
+*/
+immutable(BOMSeq) getBOM(Range)(Range input)
+        if(isInputRange!Range && is(ElementType!(Unqual!Range) == ubyte))
+{
+    import std.algorithm.searching : startsWith;
+    foreach (it; bomTable[1 .. $])
+    {
+        if (startsWith(input, it.seq))
+        {
+            return it;
+        }
+    }
+
+    return bomTable[0];
+}
+
+///
+unittest
+{
+    import std.format : format;
+
+    // This creates a dstring "Hello World" with a BOM
+    auto ts = cast(dstring)(
+        to!(ubyte[])([0xFF, 0xFE, 0x00, 0x00]) ~
+        cast(ubyte[])to!dstring("Hello World")
+    );
+    auto entry = getBOM(cast(ubyte[])ts);
+    assert(entry[0] == BOM.utf32le);
```

Comment:
`entry.bom`

(which is a bad name anyway; `BOMSeq.bom` is redundant. `BOMSeq.type` or `BOMSeq.scheme` (Unicode standard calls them Encoding Schemes) is better)